src/chain/rpc.rs

```diff
@@ -0,0 +1,724 @@
+            hashed_key_element, pallet_index, storage_key, system_balance_query,
+            system_properties_to_short_specs, unit, was_balance_received_at_account,
+        },
+    },
+    definitions::api_v2::{CurrencyProperties, OrderInfo},
+    definitions::{
+        api_v2::{AssetId, BlockNumber, CurrencyInfo, Decimals, TokenKind},
+        AssetInfo, Balance, Chain, NativeToken, Nonce, PalletIndex, Timestamp,
+    },
+    error::{Error, ErrorChain, NotHex},
+    signer::Signer,
+    state::State,
+    utils::unhex,
+    TaskTracker,
+};
+use frame_metadata::{
+    v15::{RuntimeMetadataV15, StorageEntryMetadata, StorageEntryType},
+    RuntimeMetadata,
+};
+use jsonrpsee::core::client::{ClientT, Subscription, SubscriptionClientT};
+use jsonrpsee::rpc_params;
+use jsonrpsee::ws_client::{WsClient, WsClientBuilder};
+use parity_scale_codec::{Decode, DecodeAll, Encode};
+use scale_info::{form::PortableForm, PortableRegistry, TypeDef, TypeDefPrimitive};
+use serde::{Deserialize, Deserializer};
+use serde_json::{Map, Number, Value};
+use sp_crypto_hashing::{blake2_128, blake2_256, twox_128, twox_256, twox_64};
+use std::{
+    borrow::Cow,
+    collections::{hash_map::Entry, HashMap},
+    fmt::Debug,
+    num::NonZeroU64,
+};
+use substrate_crypto_light::common::{AccountId32, AsBase58};
+use substrate_parser::{
+    cards::{Event, ExtendedData, FieldData, ParsedData, Sequence},
+    decode_all_as_type, decode_as_storage_entry,
+    special_indicators::SpecialtyUnsignedInteger,
+    storage_data::{KeyData, KeyPart},
+    AsMetadata, ShortSpecs,
+};
+
+const MAX_BLOCK_NUMBER_ERROR: &str = "block number type overflow is occurred";
+const BLOCK_NONCE_ERROR: &str = "failed to fetch an account nonce by the scanner client";
+
+const CHARGE_ASSET_TX_PAYMENT: &str = "ChargeAssetTxPayment";
+
+// Pallets
+
+const SYSTEM: &str = "System";
+const BALANCES: &str = "Balances";
+const UTILITY: &str = "Utility";
+const ASSETS: &str = "Assets";
+const BABE: &str = "Babe";
+const TRANSFER: &str = "Transfer";
+
+// Runtime APIs
+
+const AURA: &str = "AuraApi";
+
+pub async fn subscribe_blocks(client: &WsClient) -> Result<Subscription<BlockHead>, ErrorChain> {
+    Ok(client
+        .subscribe(
+            "chain_subscribeFinalizedHeads",
+            rpc_params![],
+            "chain_unsubscribeFinalizedHeads",
+        )
+        .await?)
+}
+
+pub async fn get_value_from_storage(
+    client: &WsClient,
+    whole_key: &str,
+    block: &BlockHash,
+) -> Result<Value, ErrorChain> {
+    let value: Value = client
+        .request(
+            "state_getStorage",
+            rpc_params![whole_key, block.to_string()],
+        )
+        .await?;
+    Ok(value)
+}
+
+pub async fn get_keys_from_storage(
+    client: &WsClient,
+    prefix: &str,
+    storage_name: &str,
+    block: &BlockHash,
+) -> Result<Value, ErrorChain> {
+    let keys: Value = client
+        .request(
+            "state_getKeys",
```

Comment:
state_getKeys got deprecated, suggested to use (state_getKeysPaged](https://polkadot.js.org/docs/substrate/rpc/#getkeyskey-storagekey-at-blockhash-vecstoragekey) instead

following implementation works, but can be optimised
```
pub async fn get_keys_from_storage(
    client: &WsClient,
    prefix: &str,
    storage_name: &str,
    block: &BlockHash,
) -> Result<Value, ErrorChain> {
    let storage_key_prefix = format!(
        "0x{}{}",
        hex::encode(twox_128(prefix.as_bytes())),
        hex::encode(twox_128(storage_name.as_bytes()))
    );

    let count = 100; // I expect it to always work, but will it actually?
    let start_key: Option<&str> = None; // Start from the beginning

    let mut params = vec![
        serde_json::to_value(storage_key_prefix.clone()).unwrap(),
        serde_json::to_value(count).unwrap(),
    ];

    if let Some(start_key) = start_key {
        params.push(serde_json::to_value(start_key).unwrap());
    }

    params.push(serde_json::to_value(block.to_string()).unwrap());

    let keys: Value = client
        .request(
            "state_getKeysPaged",
            params
        )
        .await
        .map_err(ErrorChain::Client)?;

    Ok(keys)
}
```